polar-core/src/data_filtering.rs

```diff
@@ -0,0 +1,799 @@
+                            field: my_field.clone(),
+                            value: ConstraintValue::Ref(Ref {
+                                field: Some(other_field.clone()),
+                                result_id: child.clone(),
+                            }),
+                        });
+                    } else {
+                        // Remove the id from the resolve_order too.
+                        result_set.resolve_order.pop();
+                    }
+                }
+
+                continue;
+            }
+            // Non relationship or unknown type info.
+            let mut contributed_constraints = false;
+            if let Some(value) = vars.eq_values.get(child) {
+                request.constraints.push(Constraint {
+                    kind: ConstraintKind::Eq,
+                    field: field.clone(),
+                    value: ConstraintValue::Term(value.clone()),
+                });
+                contributed_constraints = true;
+            }
+            if let Some(values) = vars.contained_values.get(child) {
+                for value in values {
+                    request.constraints.push(Constraint {
+                        kind: ConstraintKind::Contains,
+                        field: field.clone(),
+                        value: ConstraintValue::Term(value.clone()),
+                    });
+                }
+                contributed_constraints = true;
+            }
+            for eqf in vars
+                .field_relationships
+                .iter()
+                .filter(|r| r.0 == *parent && r.1 != *field && r.2 == *child)
+            {
+                request.constraints.push(Constraint {
+                    kind: ConstraintKind::Eq,
+                    field: field.clone(),
+                    value: ConstraintValue::Field(eqf.1.clone()),
+                });
+                contributed_constraints = true;
+            }
+            assert!(contributed_constraints);
+        }
+    }
+
+    // Constrain any vars that are `in` this var.
+    // Add their constraints to this one.
+    // @NOTE(steve): I think this is right, but I'm not totally sure.
+    // This might assume that the current var is a relationship of kind "children".
+    for (lhs, rhs) in &vars.in_relationships {
+        if rhs == var_id {
+            constrain_var(result_set, types, vars, lhs, var_type);
+            let in_result_set = result_set.requests.remove(lhs).unwrap();
+            assert_eq!(result_set.resolve_order.pop(), Some(lhs.to_string()));
+            request.constraints.extend(in_result_set.constraints);
+        }
+    }
+
+    result_set.requests.insert(var_id.to_string(), request);
+    result_set.resolve_order.push(var_id.to_string());
+}
+
+pub fn opt_pass(filter_plan: &mut FilterPlan, explain: bool) -> bool {
+    let mut optimized = false;
+
+    // Remove duplicate result set in a union.
+    let mut drop_plan = None;
```

Comment:
How does this happen? If two rules are identical?

Reply:
We get this from the builtin roles policy. Not sure why, maybe because there's like a parent check and an ancestor check?